classes/local/factor/object_factor_base.php

```diff
@@ -230,22 +230,31 @@ public function has_revoke() {
 
     /**
      * Marks factor record as revoked.
+     * If factorid is not provided, revoke all instances of factor.
      *
      * @param int $factorid
      * @return bool
      * @throws \dml_exception
      */
-    public function revoke_user_factor($factorid) {
+    public function revoke_user_factor($factorid = null) {
         global $DB, $USER;
-
-        if ($DB->set_field('factor_'.$this->name, 'revoked', 1, array('id' => $factorid))) {
-            $event = \tool_mfa\event\user_revoked_factor::user_revoked_factor_event($USER, $this->get_display_name());
-            $event->trigger();
-
-            return true;
+        if (!empty($factorid)) {
```

Comment:
Nit: this feels a little duplicated, you combine these in a $params object and then do the set_field and trigger consistently

Reply:
set_field always returns true or throws an exception so I thin the outer if is redundant